src/Diff.hs

```diff
@@ -0,0 +1,606 @@
+  where
+    both _ a b = Just (a, b)
+
+    left _ = Data.Map.empty
+
+    right _ = Data.Map.empty
+
+-- `getGroupedDiff` from `Diff` library, adapted for `patience`
+getGroupedDiff :: Ord a => [a] -> [a] -> [Patience.Item [a]]
+getGroupedDiff oldList newList = go $ Patience.diff oldList newList
+  where
+    go = \case
+      Patience.Old x : xs ->
+        let (fs, rest) = goOlds xs
+         in Patience.Old (x : fs) : go rest
+      Patience.New x : xs ->
+        let (fs, rest) = goNews xs
+         in Patience.New (x : fs) : go rest
+      Patience.Both x y : xs ->
+        let (fs, rest) = goBoth xs
+            (fxs, fys) = unzip fs
+         in Patience.Both (x : fxs) (y : fys) : go rest
+      [] -> []
+
+    goOlds = \case
+      Patience.Old x : xs ->
+        let (fs, rest) = goOlds xs
+         in (x : fs, rest)
+      xs -> ([], xs)
+
+    goNews = \case
+      Patience.New x : xs ->
+        let (fs, rest) = goNews xs
+         in (x : fs, rest)
+      xs -> ([], xs)
+
+    goBoth = \case
+      Patience.Both x y : xs ->
+        let (fs, rest) = goBoth xs
+         in ((x, y) : fs, rest)
+      xs -> ([], xs)
+
+-- | Diff two outputs
+diffOutput
+    :: Text
+    -- ^ Output name
+    -> (DerivationOutput FilePath Text)
+    -- ^ Left derivation outputs
+    -> (DerivationOutput FilePath Text)
+    -- ^ Right derivation outputs
+    -> (Maybe OutputDiff)
+diffOutput outputName leftOutput rightOutput = do
+    -- We deliberately do not include output paths or hashes in the diff since
+    -- we already expect them to differ if the inputs differ.  Instead, we focus
+    -- only displaying differing inputs.
+    let leftHashAlgo  = Nix.Derivation.hashAlgo leftOutput
+    let rightHashAlgo = Nix.Derivation.hashAlgo rightOutput
+    if leftHashAlgo == rightHashAlgo
+      then Nothing
+      else Just (OutputDiff outputName (Changed leftHashAlgo rightHashAlgo))
+
+-- | Diff two sets of outputs
+diffOutputs
+    :: Map Text (DerivationOutput FilePath Text)
+    -- ^ Left derivation outputs
+    -> Map Text (DerivationOutput FilePath Text)
+    -- ^ Right derivation outputs
+    -> OutputsDiff
+diffOutputs leftOutputs rightOutputs = do
+    let leftExtraOutputs  = Data.Map.difference leftOutputs  rightOutputs
+    let rightExtraOutputs = Data.Map.difference rightOutputs leftOutputs
+
+    let bothOutputs = innerJoin leftOutputs rightOutputs
+
+    let
+      extraOutputs =
+        if Data.Map.null leftExtraOutputs && Data.Map.null rightExtraOutputs
+          then Nothing
+          else Just (Changed leftExtraOutputs rightExtraOutputs)
+    let
+      outputDifference = flip map (Data.Map.toList bothOutputs) \(key, (leftOutput, rightOutput)) -> do
+        if leftOutput == rightOutput
+        then Nothing
+        else Just (diffOutput key leftOutput rightOutput)
+
+    OutputsDiff extraOutputs (catMaybes . catMaybes $ outputDifference)
+
+mapDiff :: (a -> b) -> Patience.Item a -> Patience.Item b
```

Comment:
Could you please check if this is different from `Functor` instance for `Item`?

Reply:
I think, we can replace it with `fmap`. This code was in the earlier version of nix-diff, I just didn't change it.